pkg/urbit/vere/pier.c

```diff
@@ -652,167 +596,136 @@ _pier_apply(u3_pier* pir_u)
+static void
+_pier_set_ship(u3_pier* pir_u, u3_noun who, u3_noun fak)
 {
-  u3_writ* wit_u;
-  c3_d     evt_d;
-  c3_l     mug_l;
-  c3_d     pos_d;
-  c3_d*    buf_d;
-  c3_d     len_d;
-
-  if ( 2 != sscanf(nam_c, "%" PRIu64 "-%x.urbit-log", &evt_d, &mug_l) ) {
-    //  fprintf(stderr, "pier: load: precommit: bad file: %s\r\n", nam_c);
-    return 0;
-  }
+  c3_assert( c3y == u3ud(who) );
+  c3_assert( (c3y == fak) || (c3n == fak) );
 
-  wit_u = c3_calloc(sizeof(*wit_u));
+  c3_o fak_o = fak;
+  c3_d who_d[2];
 
-  //  fprintf(stderr, "pier: (%" PRIu64 "): %p restore\r\n", evt_d, wit_u);
+  u3r_chubs(0, 2, who_d, who);
 
-  wit_u->pir_u = pir_u;
-  wit_u->evt_d = evt_d;
+  c3_assert( ( (0 == pir_u->fak_o) &&
+               (0 == pir_u->who_d[0]) &&
+               (0 == pir_u->who_d[1]) ) ||
+             ( (fak_o == pir_u->fak_o) &&
+               (who_d[0] == pir_u->who_d[0]) &&
+               (who_d[1] == pir_u->who_d[1]) ) );
 
-  wit_u->fol_u = u3_foil_absorb(pir_u->log_u->pre_u, nam_c);
-  if ( 0 == wit_u->fol_u ) {
-    //  fprintf(stderr, "pier: load: precommit: absorb failed: %s\r\n", nam_c);
-    c3_free(wit_u);
+  pir_u->fak_o = fak_o;
+  pir_u->who_d[0] = who_d[0];
+  pir_u->who_d[1] = who_d[1];
 
-    return 0;
-  }
-  if ( evt_d < lav_d ) {
-    // fprintf(stderr, "pier: load: precommit: already done: %s\r\n", nam_c);
-    u3_foil_delete(0, 0, wit_u->fol_u);
-    c3_free(wit_u);
+  {
+    u3_noun how = u3dc("scot", 'p', u3k(who));
 
-    return 0;
+    c3_free(pir_u->who_c);
+    pir_u->who_c = u3r_string(how);
+    u3z(how);
   }
 
-  pos_d = wit_u->fol_u->end_d;
-  if ( 0 == pos_d ) {
-    fprintf(stderr, "pier: load: precommit: empty: %s\r\n", nam_c);
-    u3_foil_delete(0, 0, wit_u->fol_u);
-    c3_free(wit_u);
-    return 0;
+  //  Disable networking for fake ships
+  //
+  if ( c3y == pir_u->fak_o ) {
+    u3_Host.ops_u.net = c3n;
   }
-  if ( 0 == (buf_d = u3_foil_reveal(wit_u->fol_u,
-                                    &pos_d,
-                                    &len_d)) )
+
+  u3z(who); u3z(fak);
+}
+
+/* _pier_disk_read_header_complete():
+** XX async
+*/
+static void
+_pier_disk_read_header_complete(u3_disk* log_u, u3_noun dat)
+{
+  u3_pier* pir_u = log_u->pir_u;
+
   {
-    //  fprintf(stderr, "pier: load: precommit: reveal failed: %s\r\n", nam_c);
-    u3_foil_delete(0, 0, wit_u->fol_u);
-    c3_free(wit_u);
+    u3_noun who, fak, len;
 
-    return 0;
-  }
-  wit_u->mat = u3i_chubs(len_d, buf_d);
-  c3_free(buf_d);
+    u3x_trel(dat, &who, &fak, &len);
 
-  if ( mug_l != u3r_mug(wit_u->mat) ) {
-    //  fprintf(stderr, "pier: load: precommit: reveal failed: %s\r\n", nam_c);
-    u3_foil_delete(0, 0, wit_u->fol_u);
-    u3z(wit_u->mat);
-    c3_free(wit_u);
+    c3_assert( c3y == u3ud(who) );
+    c3_assert( 1 >= u3r_met(7, who) );
+    c3_assert( c3y == u3ud(fak) );
+    c3_assert( 1 >= u3r_met(0, fak) );
+    c3_assert( c3y == u3ud(len) );
+    c3_assert( 1 >= u3r_met(3, len) );
```

Comment:
Does this mean the boot sequence can't be more than seven events long?

Reply:
Can't be more than 255.

Also, this specifically is just the length of the lifecycle sequence, the list of nock formulas in the head of the pill that must be evaluated via the lifecycle formula `[2 [0 3] 0 2]`. It's not the full boot sequence.